PPA: Collect Client Metadata (EXPOSUREAPP-5147)

kolyaopahle opened the pull request:
### Description

This PR implements the Client Metadata Donor for the Privacy Preserving Analytics

### Steps to reproduce
1. Do an analytics collection in the test menu
2. Check if submitted data matches app / device specs

kolyaopahle committed:
Implemented ClientMetadataDonor which uses ClientVersionParser to extract the major minor and patch version from the app VERSION_CODE

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Restructured ClientMetadataDonor and added ClientMetadataDonorTests

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Fixed comment in ClientMetadataDonor

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/clientmetadata/ClientMetadataDonor.kt:
```diff
@@ -0,0 +1,52 @@
+) : DonorModule {
+
+    override suspend fun beginDonation(request: DonorModule.Request): DonorModule.Contribution {
+        val config = appConfigProvider.getAppConfig()
```
We should use `configProvider.latestConfig` here. I don't think a 5 hour old config would be a problem here and `Analytics` will already attempt a refresh when checking for conditions.

kolyaopahle replied:
Sounds good 👍

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/clientmetadata/ClientVersionWrapper.kt:
```diff
@@ -0,0 +1,12 @@
+import javax.inject.Inject
+
+@Reusable
+class ClientVersionWrapper @Inject constructor() {
```
Do we need an extra class for this?

We also already have `ApiLevel`.

kolyaopahle replied:
ah nice, didn't know about ApiLevel 👍

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/clientmetadata/ClientVersionParser.kt:
```diff
@@ -0,0 +1,37 @@
+@Reusable
+class ClientVersionParser @Inject constructor() {
+    fun parseClientVersion(buildNumber: Int): ClientVersion {
+        val major = buildNumber / MAJOR_OFFSET
```
Nice

d4rken replied:
If we change the way the version is calculated this needs to change too. Please add a test case that fails if the buildNumber pattern changes, e.g. gets larger for different offsets.

kolyaopahle replied:
makes sense, added a test that compares a generated version string with the original version string 👍

d4rken reviewed (changes requested):
Code works good, would like a few small additions :hugs:

kolyaopahle committed:
Got rid of ClientVersionWrapper, switched all usages to ApiLevel
Added ability to generate version string to ClientVersionParser
Added ClientVersionParserTest that compares generated string with real one to make sure the parser is working
Changed ClientMetadataDonor to use currentConfig instead of getAppConfig

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

d4rken commented on Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/modules/clientmetadata/ClientVersionParserTest.kt:
```diff
@@ -0,0 +1,35 @@
+
+        val parsedVersion = createInstance().parseClientVersion(currentVersionNumber)
+
+        parsedVersion.toVersionString() shouldBe currentVersionName
```
Careful. Only `deviceForTesters` Variant has `-RCx`, would this unit test fail  if executed for a different build flavor?

kolyaopahle replied:
noted, tossed out the whole parsing of version codes and switched to major minor and patch version being provided by the buildconfig

kolyaopahle committed:
Removed ClientVersionParser and put MAJOR, MINOR and PATCH version directly in the BuildConfig

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Merge branch 'release/1.14.x' into feature/5147-collect-client-metadata

d4rken reviewed (approved):
lgtm

chiljamgossow reviewed (approved):
Tested, LGTM

kolyaopahle committed:
Merge branch 'release/1.14.x' into feature/5147-collect-client-metadata
